tsdb/chunks/chunks.go

```diff
@@ -284,27 +301,75 @@ func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
 	return newChunk, nil
 }
 
+// WriteChunks writes as many chunks as possible to the current segment,
+// cuts a new segment when the current segment is full and
+// writes the rest of the chunks in the new segment.
 func (w *Writer) WriteChunks(chks ...Meta) error {
-	// Calculate maximum space we need and cut a new segment in case
-	// we don't fit into the current one.
-	maxLen := int64(binary.MaxVarintLen32) // The number of chunks.
-	for _, c := range chks {
-		maxLen += binary.MaxVarintLen32 + 1 // The number of bytes in the chunk and its encoding.
-		maxLen += int64(len(c.Chunk.Bytes()))
-		maxLen += 4 // The 4 bytes of crc32
-	}
-	newsz := w.n + maxLen
+	var (
+		chksBatchSize int64
+		end           int
+	)
 
-	if w.wbuf == nil || newsz > w.segmentSize && maxLen <= w.segmentSize {
+	// w.wbuf == nil means it is the first chunk
+	// so need to start a new segment.
+	if w.wbuf == nil {
 		if err := w.cut(); err != nil {
 			return err
 		}
 	}
 
+	for _, chk := range chks {
+		// Each chunk contains: data length + encoding + the data itself + crc32
+		chksBatchSize += int64(MaxChunkLengthFieldSize) // The data length is a variable length field so use the maximum possible value.
+		chksBatchSize += ChunkEncodingSize              // The chunk encoding.
+		chksBatchSize += int64(len(chk.Chunk.Bytes()))  // The data itself.
+		chksBatchSize += crc32.Size                     // The 4 bytes of crc32
+
+		end++
+		if chksBatchSize+w.n > w.segmentSize {
+			if end > 1 {
+				// Don't include the last chunk only if there are >1 chunks.
+				// This will keep segment size within the configured limit.
+				// If a single chunks is bigger than the configured limit,
+				// we cannot do much.
+				end--
```

Comment:
Because we try to keep the segment size within the configured limit.  
So, if `end == 1` and if the current segment has chunks, then cut a new segment for new chunks.

Reply:
sorry I didn't understand what you mean.

Reply:
sorry, I will try to make it clear. (:

Reply:
For example:    
    1. `w.segmentSize`: 1000   // configured limit
    2. `w.n`:  950                         // segment offset, there are some chunks in the current segment 
    3. the size of a new chunk:  120  

should write the new chunk to the current segment, or write to a new segment?

Reply:
If our goal is to keep the segment size within the configured limit, the new chunk should be written to a new segment in this example.  right?

Reply:
Yep you are right and I have added tests to cover this condition as well and unfortunately this adds a lot more logic to the implementation. I don't like the implementation right now as it is too complex, so will look into it in few days with some fresh perspective. Please feel free to comment if you can see a way to simplify it and still pass all tests.

If we can't find a way to simplify the current implementation I would prefer to be not so strict about the segment size if this makes the implementation easier to understand.

Reply:
Yeah it makes the implementation more difficult to understand.

> If we can't find a way to simplify the current implementation I would prefer to be not so strict about the segment size if this makes the implementation easier to understand.

Fully agree with you.

Reply:
How do you think about this way?
```
		if chksBatchSize+w.n > w.segmentSize {
			if end == 1 && w.n > SegmentHeaderSize {
				if err := w.cut(); err != nil {
					return err
				}
				end--
			} else if end > 1 {
				end--
			}
			// . . .

```

https://github.com/zhulongcheng/prometheus/commit/7700541485589c65b768c54bc241917ba3b34426

Reply:
I just tried it and the tests fail, do the test pass at your end?

Reply:
I noticed that we can't reset the `chks` inside for loop.